archunit/src/main/java/com/tngtech/archunit/core/importer/AccessRecord.java

```diff
@@ -273,17 +282,59 @@ private static JavaCodeUnit getCaller(CodeUnit caller, ImportedClasses classes)
                     " that matches supposed caller " + caller);
         }
 
-        private static <T extends HasName & HasDescriptor & HasOwner<JavaClass>> Set<T>
-        tryFindMatchingTargets(Set<T> possibleTargets, TargetInfo targetInfo) {
-            ImmutableSet.Builder<T> result = ImmutableSet.builder();
-            for (T possibleTarget : possibleTargets) {
-                if (targetInfo.matches(possibleTarget)) {
+        private static <MEMBER extends JavaMember, TARGET extends TargetInfo> Set<MEMBER>
+        tryFindMatchingTargets(Set<MEMBER> possibleTargets, TARGET target, SignaturePredicate<TARGET> signaturePredicate) {
+            ImmutableSet.Builder<MEMBER> result = ImmutableSet.builder();
+            for (MEMBER possibleTarget : possibleTargets) {
+                if (matches(possibleTarget, target, signaturePredicate)) {
                     result.add(possibleTarget);
                 }
             }
             return result.build();
         }
 
+        private static <MEMBER extends JavaMember, TARGET extends TargetInfo> boolean matches(MEMBER member, TARGET target, SignaturePredicate<TARGET> signaturePredicate) {
+            if (!target.name.equals(member.getName()) || !target.desc.equals(member.getDescriptor())) {
+                return false;
+            }
+            return target.owner.getFullyQualifiedClassName().equals(member.getOwner().getName()) ||
+                    containsExactlyOneMatch(new ClassHierarchyPath(target.owner, member.getOwner()), target, signaturePredicate);
+        }
+
+        private static <TARGET extends TargetInfo> boolean containsExactlyOneMatch(Iterable<JavaClass> classes, TARGET target, SignaturePredicate<TARGET> signaturePredicate) {
+            Set<JavaClass> matching = new HashSet<>();
+            for (JavaClass javaClass : classes) {
+                if (signaturePredicate.exists(javaClass, target)) {
+                    matching.add(javaClass);
+                }
```

Comment:
Yeah, this only occurs, if a method has been overridden several times in the hierarchy. E.g. if you have

```
class Base {
  void foo() {
    // some impl
  }
}

class Child extends Base {
  @Override
  void foo() {
    // also some impl
  }
}

class GrandChild extends Child {
  // not overridden here, otherwise it would be directly found
}
```

And then by accident you pick `Base.foo()` as the first candidate to find the target of a call to `GrandChild.foo()`. I'll still look into it :+1: (unfortunately the whole "start with any method in the hierarchy is pretty sub-optimal" :disappointed:)

Reply:
Hmm, I think I'll leave it out for now. I could not find any noticeable difference in performance. And the problem is, that the benefit is not completely clear for me. We are exchanging a couple of unnecessary loops in a rare case against an operation we have to execute in every single loop. Take the extreme case that we for example have no single method that has been overridden multiple times in the whole set of imported classes, then we would actually even perform worse than if we don't do the check.

Reply:
I absolutely agree.
When I initially thought about it, I hadn't realized what `containsExactlyOneMatch` is actually used for...